native forth word contributions

patricksurry opened the pull request:
This is a non-functional change, just correcting a couple of typos, and contributing a bunch of native words I've been collecting.  These are not included in the build but could be pulled in by the platform files.  If you prefer some other folder structure for these words let me know.

I want to try migrating my own hardware support as a proper platform after your re-org.   

I have a bunch of kernel code that evolves with my hardware, which I'm not sure should live in the taliforth repo?   Perhaps I'd still have my own repo that is required to build my 'platform'?  I'll try a few things and see what makes sense.  But let me know if you have ideas on that.

patricksurry committed:
native forth word contributions

SamCoVT commented:
Oh good - I was hoping to get a chance to see your `BIND` word.
I think these are great.  Your modified DUMP might be a good candidate for the neo6502 platform as it only has a 53 character wide screen.

I don't think I like them in the words folder (but you could probably talk me into it).  They don't make much sense in the forth_examples folder as that's really for Forth code.  Does it make sense to make an "assembly_examples" or "native_examples" folder?  ...or to make an "examples" folder with a Forth subfolder and an assembly subfolder?  They kind of make sense in the platform folder, but I don't think I want to confuse new users with both `skeleton` and `examples` in there - a subfolder under the skeleton folder might be OK though.  Let me know if you have any suggestions/thoughts.

From a platform file, they can either be included wherever they live or they can be copied into a platform folder and included there, or the code can be copied directly into the platform.asm file.  I may have to think about the Makefile a bit more - it would be nice for all .asm files in the platform folder to be dependencies; that way, someone could just copy these into their platform folder, `.include` them, and get a proper rebuild to their binary if they edit the assembly file.  I think I can do that with `wildcard` combined with `%`.

One extra thought, and it has to do with your `dump` word - should users be able to "override" Tali's inbuilt word by using the same labels?  We could, in the example of dump, make the xt_dump, w_dump, and z_dump labels weak.  I don't know if it makes sense to do that for all of the words.  The header (and flags) would need to remain the same because I don't think there is an easy way to override the header itself.  If that's something that needs a lot of discussion, we can open a separate issue for it.

patricksurry commented:
In terms of directory structure, I like the idea of a top-level `examples` or `contrib` folder to emphasize that the content there is not necessarily of the same quality/level of testing as the core stuff.   I don't think it needs to be in platform since they aren't platform-specific.   Perhaps examples/forth and examples/native or assembly would work?   

It would be nice to get rid of the forth_code folder at the same time .

I have a few more words which are platform-specific for my h/w but I would keep those ones directly in my platform folder.

Having flexibility to replace core words would be cool; I do that currently with manual .if / .endif around large words that I don't need, with more granularity than the existing groups of words, e.g. replacing DUMP with DMP and removing DISASM which I don't currently have room for.   My ideal would be a TALIFORTH_EXCLUDE = ["dump", "disasm", ...] but that seems like it would need a ton of ugly conditional logic in the source.

For DUMP in particular another approach would be merging the two versions in words/tools with some MAX_LINE_LENGTH parameter that decided which variant to compile.  (If you like the feature of always labeling by multiples of line length I could also port that back into the wider version.)

SamCoVT commented:
I like `examples`.
If you want to remove the forth_code folder (now is a good time to do it while we are shuffling everything else around), I think you need to:

- [ ] Move the `forth_to_ophisbin.py` tool in there to `tools` (and rename it, because we don't use ophis as the assembler anymore)
- [ ] Update the makefile in the places where it calls that tool.
- [ ] Get rid of the `user_words.fs` file completely.
- [ ] Consider getting rid of `forth_words.fs` completely, or move it somewhere else
    - [ ] If you moved it, update the Makefile to look for the source there (you can leave the resulting .asc file in the Tali root directory)
    - [ ] If you remove it, perhaps just put the startup strings directly into the taliforth.asm source code (see next item)
- [ ] Update `taliforth.asm` where it brings in those files.  Note that this currently has a check (`.if forth_words_start == *`) that checks to see if the weak label for forth_words_start was overridden (in a platform file) - if so, it does not include the .asc files with `.binary`.  You should determine what you want to do in that entire section from `.weak` to `.endweak` (Search for `Bring in the words` to find that section)
- [ ] (edited to add this item) Perhaps get rid of the forth_words_start/end labels and just use the user_words_start/end labels.
    - [ ] You will need to edit COLD as it uses user_words_end and forth_words_start to grab both of those as a single string.
    - [ ] You can edit the `platform/skeleton/platform.asm` file as it has both sets of labels (so that COLD could find them).

I'm not sure if that's all the details - it's probably worth checking the manual for mentions of those .fs files and just update it for whatever is put in place.

patricksurry committed:
reorganize

patricksurry commented:
I had a go at reorganizing:

- I got rid of forth_words completely, moving the default splash text into taliforth.asm.  Any platform forth replaces this.
- There are some remaining Ophis references in various places.  I removed the ones related to the (renamed) `forth_pack.py`
- I refactored `forth_pack.py` slightly so I could make `test_forth_pack.py` work again.  (It had rusted.)
- I moved the assembly listing and labels into the platform folder instead of docs which always seemed weird to me (is it needed there for some reason?)
- I made some changes to README and doc files but might need more

patricksurry committed:
revert results

SamCoVT commented:
I agree with listings in the platform folder now.
When Tali was being mainly developed on the py65mon simulator, perhaps having the listing in with the other docs made sense; now that's it's spread to other hardware (real and virtual), I think putting the listing and labelmap in the platform folder makes the most sense.

Oh, and thanks for the gitignore of the bin files.  I was starting to grow quite a collection of those as I was making sure all of the other platforms still assembled properly.

I think this looks pretty good.  Let me know once you are happy with the documentation and I''ll merge it.

patricksurry committed:
docs

patricksurry committed:
revert some unneeded stuff

patricksurry commented:
i did a few more doc tweaks, I think it's in decent shape
